compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td

```diff
@@ -1506,4 +1506,63 @@ def IREELinalgExt_WinogradOutputTransformOp : IREELinalgExt_Op<"winograd.output_
 
 } // OpGroupWinogradOps
 
+//===---------------------------------------------------------------------===//
+// Custom tilable op
+//===---------------------------------------------------------------------===//
+
+def IREELinalgExt_CustomOp : IREELinalgExt_Op<"custom_op"> {
+  let summary = "Custom operation for compiling with IREE";
+  let description = [{
+    This operation is meant to allow computation sequences that are fused at
+    tile level prescriptively. This is to account for cases where such fusion
+    cannot/is not yet discovered appropriately.
+
+    The operation implements all the interfaces needed to be able to
+    1. Compile e2e using IREE
+    2. Still be able to fuse with other operations that the compiler can
+       figure out automatically.
+
+    The operation has operands similar to `LinalgOp`s
+    1. Indexing maps : Modification on top of Linalg Op is that the indexing
+       map can be `affine_map<() -> ()>` indicating that operand cannot
+       be fused along and that this operand is not tiled
```

Comment:
I feel that we need more control than per-operand. Take a similar problem to attention that motivated the addition of this op.

```
iree_linalg_ext.custom_op ins(%q: tensor<?x?>, %k: tensor<?x?>, %v: tensor<?x?>) {
  ^bb0(%q_tile: tensor<?x?>, %k_tile: tensor<?x?>, %v_tile: tensor<?x?>)
  %0 = linalg.matmul_transpose_b ins(%q_tile, %k_tile)
  %1 = linalg.matmul_transpose_b ins(%0, %v_tile)
  linalg_ext.yield %1
}
```

The question here is how to design the indexing maps for this op. This op effectively has `4` loops hidden inside of it, the `M` and `N` loops of the second matmul, and the reduction loop for each matmul.

```
iterator_types = [parallel, parallel, reduction, reduction]
```

There is an issue here though, where the first reduction loop is not tilable as a part of this operation. Tiling the first reduction loop would need to produce a loop around only the first matmul, which is incompatible with the tiling interface which forms the loop then invokes the interface on a clone of the full operation.

So if we can't tile that reduction loop, how do we design affine maps for the `Q` operand. Well the `Q` operation is shaped `tensor<M x K1>`, which is a tilable parallel loop and untilable dimension so there isn't a clear answer to what the indexing map should be.

I think we should just make the iteration space on the op explicit to mitigate issues like this. The status of the `K1` loop as a reduction loop would be recovered after decomposing (inlining) this custom op, at which point we can do further lowerings/tilings if needed.

Reply:
I also had the same concern which I pointed out to Mahesh. One suggestion I had was representing these non-tilable dimensions with a special iterator type that explicitly states that these dimensions are non-tilable. These dimensions would also need to be the inner-most dimensions of the operation, otherwise you cannot tile any dimension that occurs after these non-tilable dimensions (you cannot move a loop out of a non-tilable loop, because it is non-perfectly nested).

I do think that there is something hacky going on here. I'm going to write out more examples related to why I think there will be problems with attention and hopefully present them next week when we sync.

Reply:
Based on discussions I changed it to use indexing maps with symbols to represent accesses to tensors which are not done through the "iterators" of the op. I am not sure if that works fully, but will adapt based on implementation.